### PR TITLE
[receiver/gelfreceiver] Added Support for GELF log receiver.

### PR DESCRIPTION
**Pull request by BharatKJain**
**Description:** <Describe what has changed.>

Added support for GELF based logs - [please read this document to understand GELF implementation](https://archivedocs.graylog.org/en/latest/pages/gelf.html).

[Tremor's implementation for GELF receiver
](https://www.tremor.rs/docs/0.12/reference/preprocessors/gelf-chunking/)

Otel-collector's GELF implementation document: https://docs.google.com/document/d/1P9JcrXrORqI0wsnRgrcUroluutf3rVrcUCIgLeSpFm0/edit?usp=sharing

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/33861

**Testing:** 

- Tested the functionality of the code locally.
- Will create the unit-tests for it.

**Documentation:** <Describe the documentation added.> Work-in-progress.

### CODE REVIEW

**Review by djaglowski — CHANGES_REQUESTED** [review]
In order to accept a new component you must [formally propose it in an issue](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/new?assignees=&labels=Sponsor+Needed%2Cneeds+triage&projects=&template=new_component.yaml&title=New+component%3A+) and find a sponsor.

**Comment by BharatKJain** [comment]
> In order to accept a new component you must [formally propose it in an issue](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/new?assignees=&labels=Sponsor+Needed%2Cneeds+triage&projects=&template=new_component.yaml&title=New+component%3A+) and find a sponsor.

@djaglowski I don't have a sponsor, is it possible for you to help with that part? or is there a process to get sponsor, then please share the readme/doc for it?

**Comment by djaglowski** [comment]
I don't have capacity to voluntarily sponsor any additional components. Hopefully someone else is interested. Otherwise, you can always host the component elsewhere and others can still build it into their collectors using the Collector Builder.

**Comment by BharatKJain** [comment]
@djaglowski Instead of creating a separate receiver, if I add this as a part of UDP receiver then will that be okay? It will be like a CODEC.

**Comment by djaglowski** [comment]
I don't believe GELF support should be a concern of the UDP receiver.

**Review by hughesjj — CHANGES_REQUESTED** [review]
I don't see any tests, could you please add some to ensure the receiver works as expected?

**Comment by BharatKJain** [comment]
> I don't see any tests, could you please add some to ensure the receiver works as expected?

@hughesjj I tried talking to folks in SIG otel-collector, they said that we need sponsor for this PR and that it won't be accepted unless there's someone sponsoring this, so before I spend more effort on it - wanted to know if you would be able to sponsor this PR or get this accepted given that all comments are fixed?

I don't want my efforts to go into vain. 😞

File: receiver/gelfreceiver/docs.go [hunk 1]
```diff
@@ -0,0 +1,3 @@
+//go:generate mdatagen metadata.yaml
```
Comment (atoulme) [inline receiver/gelfreceiver/docs.go:1]: copyright header is missing

File: cmd/otelcontribcol/go.mod [hunk 2]
```diff
@@ -2,9 +2,7 @@
 
 module github.com/open-telemetry/opentelemetry-collector-contrib/cmd/otelcontribcol
 
-go 1.21.0
-
-toolchain go1.21.11
+go 1.22.0
```
Comment (atoulme) [inline cmd/otelcontribcol/go.mod:5]: please revert

File: cmd/otelcontribcol/go.mod [hunk 3]
```diff
@@ -141,6 +139,7 @@ require (
 	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/filestatsreceiver v0.102.0
 	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/flinkmetricsreceiver v0.102.0
 	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/fluentforwardreceiver v0.102.0
+	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/gelfreceiver v0.0.0
```
[the remark this answers is not in the document]
Comment (atoulme) [reply cmd/otelcontribcol/go.mod:142]: yes, and same in the builder yaml. If anything, you should defer adding this receiver to cmd/otelcontribcol to a second PR.

File: pkg/stanza/operator/input/gelfinternal/config.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+				return &buffer
+			},
+		},
+		buffer:          make(map[string]*MapGelfMessage),
+		lastBuffer:      make(map[string]*MapGelfMessage),
+		asyncReaders:    c.AsyncReaders,
+		asyncProcessors: c.AsyncProcessors,
+	}
+
+	// fmt.Println("GELF receiver config validated.")
```
Comment (atoulme) [inline pkg/stanza/operator/input/gelfinternal/config.go:110]:
```suggestion
```

File: pkg/stanza/operator/input/gelfinternal/config.go [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+// Build will build a udp input operator.
+func (c Config) Build(set component.TelemetrySettings) (operator.Operator, error) {
+	inputOperator, err := c.InputConfig.Build(set)
+	if err != nil {
+		return nil, err
+	}
+
+	if c.ListenAddress == "" {
+		return nil, fmt.Errorf("missing required parameter 'listen_address'")
```
Comment (atoulme) [inline pkg/stanza/operator/input/gelfinternal/config.go:73]:
```suggestion
		return nil, errors.New("missing required parameter 'listen_address'")
```

File: pkg/stanza/operator/input/gelfinternal/config.go [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+	}
+
+	if _, _, err := net.SplitHostPort(c.ListenAddress); err != nil {
+		return nil, fmt.Errorf("invalid listen_address: %w", err)
+	}
+
+	if c.Protocol != "udp" {
+		return nil, fmt.Errorf("supported protocols - udp, invalid protocol: %s", c.Protocol)
+	}
+	if c.AsyncReaders <= 0 && c.AsyncReaders < 20 {
```
Comment (atoulme) [inline pkg/stanza/operator/input/gelfinternal/config.go:83]: are you sure of these conditions?

File: pkg/stanza/operator/input/gelfinternal/config.go [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+		return nil, fmt.Errorf("invalid async_reader: %d", c.AsyncReaders)
+	}
+	if c.AsyncProcessors <= 0 && c.AsyncProcessors < 20 {
+		return nil, fmt.Errorf("invalid async_processors: %d", c.AsyncProcessors)
+	}
+	if c.UDPMaxQueueLength <= 0 && c.UDPMaxQueueLength < 65535 {
```
Comment (atoulme) [inline pkg/stanza/operator/input/gelfinternal/config.go:89]:
```suggestion
	if c.UDPMaxQueueLength <= 0 || c.UDPMaxQueueLength > 65535 {
```

File: pkg/stanza/operator/input/gelfinternal/config.go (hunk 6 above)
Comment (BharatKJain) [reply pkg/stanza/operator/input/gelfinternal/config.go:83]: I am not sure, just wanted to add some boundary conditions, I'll remove it.

File: pkg/stanza/operator/input/gelfinternal/config.go (hunk 5 above)
Comment (BharatKJain) [reply pkg/stanza/operator/input/gelfinternal/config.go:73]: I am not sure but if you check ../udp/config.go, they are using fmt...I didn't wanted to diverge from it, let me know your thoughts.

File: receiver/gelfreceiver/docs.go (hunk 1 above)
Comment (BharatKJain) [reply receiver/gelfreceiver/docs.go:4]: fixed

**Comment by BharatKJain** [comment]
- Added the suggested fixes, please re-check once.
- Added unit-test, will add more in sometime

Let me know if there's anything else!

**Pull request closed without merging**